File: golf_bits/lib/models/tee_label_normalize.dart
// Repeatable provider tee_name → display label + filter metadata (GolfCourseAPI / USGA-style).

/// Parsed metadata from a provider `tee_name` (or legacy DB label).
class ParsedTeeMetadata {
  const ParsedTeeMetadata({
    required this.rawLabel,
    this.color,
    this.gender,
    this.isHistorical = false,
    this.isTemporary = false,
    this.isNineHoleHint = false,
  });

  final String rawLabel;
  final String? color;
  /// `men` or `women` when detected.
  final String? gender;
  final bool isHistorical;
  final bool isTemporary;
  final bool isNineHoleHint;

  /// Stable key for picking one tee per color/gender family.
  String get familyKey {
    final c = (color ?? rawLabel.trim().toLowerCase()).toLowerCase();
    final g = gender ?? '';
    return '$c|$g';
  }
}

/// Longest-first so "Terra-Cotta" wins over shorter substring matches.
const _knownColorTokens = <String>[
  'terra-cotta',
  'terra cotta',
  'championship',
  'burgundy',
  'maroon',
  'purple',
  'silver',
  'golden',
  'yellow',
  'orange',
  'bronze',
  'cream',
  'shark',
  'ozzie',
  'aussie',
  'royal',
  'combo',
  'lion',
  'black',
  'blue',
  'white',
  'green',
  'gold',
  'red',
];

final _historicalToken = RegExp(
  r'\b(?:NOV|JAN|FEB|MAR|APR|MAY|JUN|JUL|AUG|SEP|OCT|DEC)\d{2}\b',
  caseSensitive: false,
);
final _yearSuffix = RegExp(r'\b(19|20)\d{2}\b');
final _nineHoleHint = RegExp(r'\b(?:SP\s*)?9\s*holes?\b', caseSensitive: false);
final _numericCourseCode = RegExp(r'^\d{4,6}$');
final _ratingSystem = RegExp(r'^(USGA|GA|MGA|AGU)$', caseSensitive: false);

/// Parse provider-style or plain tee names into filter/display metadata.
ParsedTeeMetadata parseProviderTeeLabel(String rawLabel) {
  final raw = rawLabel.trim();
  if (raw.isEmpty) {
    return ParsedTeeMetadata(rawLabel: raw);
  }

  final lower = raw.toLowerCase();
  final isHistorical = _historicalToken.hasMatch(raw) ||
      (_yearSuffix.hasMatch(raw) && raw.contains(','));
  final isTemporary = lower.contains('temp');
  final isNineHoleHint = _nineHoleHint.hasMatch(raw);

  if (!raw.contains(',')) {
    final color = _matchColorToken(raw) ?? _titleCaseSingle(raw);
    return ParsedTeeMetadata(
      rawLabel: raw,
      color: color,
      isHistorical: isHistorical,
      isTemporary: isTemporary,
      isNineHoleHint: isNineHoleHint,
    );
  }

  final parts = raw.split(',').map((p) => p.trim()).where((p) => p.isNotEmpty).toList();
  String? gender;
  String? color;

  for (final part in parts) {
    final pl = part.toLowerCase();
    if (_numericCourseCode.hasMatch(part) || _ratingSystem.hasMatch(part)) continue;
    if (pl == 'men' || pl == 'man' || pl == 'male' || pl == 'mens') {
      gender = 'men';
      continue;
    }
    if (pl == 'women' || pl == 'woman' || pl == 'female' || pl == 'ladies' || pl == 'womens') {
      gender = 'women';
      continue;
    }
    final matched = _matchColorToken(part);
    if (matched != null) {
      color ??= matched;
    }
  }

  color ??= _matchColorToken(raw);

  return ParsedTeeMetadata(
    rawLabel: raw,
    color: color,
    gender: gender,
    isHistorical: isHistorical,
    isTemporary: isTemporary,
    isNineHoleHint: isNineHoleHint,
  );
}

String? _matchColorToken(String text) {
  final lower = text.toLowerCase();
  for (final token in _knownColorTokens) {
    if (lower.contains(token)) {
      return _titleCaseToken(token);
    }
  }
  if (lower.contains('/')) {
    final names = <String>[];
    for (final seg in lower.split('/')) {
      final s = seg.trim();
      for (final token in _knownColorTokens) {
        if (s.contains(token)) {
          names.add(_titleCaseToken(token));
          break;
        }
      }
    }
    if (names.isNotEmpty) return names.join('/');
  }
  return null;
}

String _titleCaseToken(String token) {
  if (token.contains('-')) {
    return token.split('-').map(_titleCaseSingle).join('-');
  }
  if (token.contains(' ')) {
    return token.split(' ').map(_titleCaseSingle).join(' ');
  }
  return _titleCaseSingle(token);
}

String _titleCaseSingle(String s) {
  if (s.isEmpty) return s;
  return s[0].toUpperCase() + s.substring(1).toLowerCase();
}

/// User-facing primary label (yardage / ratings stay in subtitle).
String teeDisplayLabel(
  ParsedTeeMetadata meta, {
  required bool showGenderSuffix,
}) {
  final base = meta.color ?? _titleCaseSingle(meta.rawLabel.split(',').first.trim());
  if (showGenderSuffix && meta.gender != null) {
    final g = meta.gender == 'women' ? 'Women' : 'Men';
    return '$base ($g)';
  }
  return base;
}

/// Color hint for tee row styling (falls back to parsed color).
String? teeColorHintFromMetadata(ParsedTeeMetadata meta, String? existingHint) {
  if (meta.color != null) return meta.color;
  final h = existingHint?.trim();
  if (h != null && h.isNotEmpty) return h;
  return null;
}

int teeColorSortRank(String? color) {
  final c = (color ?? '').toLowerCase();
  const order = [
    'black',
    'championship',
    'blue',
    'white',
    'yellow',
    'red',
    'gold',
    'green',
    'silver',
    'burgundy',
    'orange',
    'purple',
    'cream',
    'bronze',
    'terra-cotta',
    'terra cotta',
  ];
  for (var i = 0; i < order.length; i++) {
    if (c.contains(order[i])) return i;
  }
  return 50;
}

/// Whether [candidate] should replace [incumbent] within the same family.
bool preferTeeCandidate({
  required bool incumbentHas18,
  required bool candidateHas18,
  required ParsedTeeMetadata incumbentMeta,
  required ParsedTeeMetadata candidateMeta,
  required int incumbentYardage,
  required int candidateYardage,
  required int incumbentHoleRows,
  required int candidateHoleRows,
}) {
  if (incumbentHas18 != candidateHas18) return candidateHas18;
  if (incumbentMeta.isHistorical != candidateMeta.isHistorical) {
    return !candidateMeta.isHistorical;
  }
  if (incumbentMeta.isTemporary != candidateMeta.isTemporary) {
    return !candidateMeta.isTemporary;
  }
  if (incumbentMeta.isNineHoleHint != candidateMeta.isNineHoleHint) {
    return !candidateMeta.isNineHoleHint;
  }
  if (incumbentYardage != candidateYardage) return candidateYardage > incumbentYardage;
  if (incumbentHoleRows != candidateHoleRows) return candidateHoleRows > incumbentHoleRows;
  return candidateMeta.rawLabel.length < incumbentMeta.rawLabel.length;
}

int compareTeeMetadataForDisplay(ParsedTeeMetadata a, ParsedTeeMetadata b) {
  final ga = a.gender == 'men' ? 0 : (a.gender == 'women' ? 1 : 2);
  final gb = b.gender == 'men' ? 0 : (b.gender == 'women' ? 1 : 2);
  if (ga != gb) return ga.compareTo(gb);
  return teeColorSortRank(a.color).compareTo(teeColorSortRank(b.color));
}

/// Same rules as [parseProviderTeeLabel] for Edge Function ingest (TypeScript port).
String normalizeProviderTeeLabelForStorage(String rawLabel) {
  final meta = parseProviderTeeLabel(rawLabel);
  return teeDisplayLabel(meta, showGenderSuffix: meta.gender != null);
}

File: supabase/functions/_shared/tee-label-normalize.ts
// Repeatable provider tee_name rules — keep in sync with golf_bits/lib/models/tee_label_normalize.dart

export type ParsedTeeMetadata = {
  rawLabel: string;
  color: string | null;
  gender: 'men' | 'women' | null;
  isHistorical: boolean;
  isTemporary: boolean;
  isNineHoleHint: boolean;
  familyKey: string;
};

const KNOWN_COLOR_TOKENS = [
  'terra-cotta',
  'terra cotta',
  'championship',
  'burgundy',
  'maroon',
  'purple',
  'silver',
  'golden',
  'yellow',
  'orange',
  'bronze',
  'cream',
  'shark',
  'ozzie',
  'aussie',
  'royal',
  'combo',
  'lion',
  'black',
  'blue',
  'white',
  'green',
  'gold',
  'red',
] as const;

const HISTORICAL_TOKEN = /\b(?:NOV|JAN|FEB|MAR|APR|MAY|JUN|JUL|AUG|SEP|OCT|DEC)\d{2}\b/i;
const YEAR_SUFFIX = /\b(19|20)\d{2}\b/;
const NINE_HOLE_HINT = /\b(?:SP\s*)?9\s*holes?\b/i;
const NUMERIC_COURSE_CODE = /^\d{4,6}$/;
const RATING_SYSTEM = /^(USGA|GA|MGA|AGU)$/i;

function titleCaseSingle(s: string): string {
  if (!s) return s;
  return s[0].toUpperCase() + s.slice(1).toLowerCase();
}

function titleCaseToken(token: string): string {
  if (token.includes('-')) return token.split('-').map(titleCaseSingle).join('-');
  if (token.includes(' ')) return token.split(' ').map(titleCaseSingle).join(' ');
  return titleCaseSingle(token);
}

function matchColorToken(text: string): string | null {
  const lower = text.toLowerCase();
  for (const token of KNOWN_COLOR_TOKENS) {
    if (lower.includes(token)) return titleCaseToken(token);
  }
  if (lower.includes('/')) {
    const names: string[] = [];
    for (const seg of lower.split('/')) {
      const s = seg.trim();
      for (const token of KNOWN_COLOR_TOKENS) {
        if (s.includes(token)) {
          names.push(titleCaseToken(token));
          break;
        }
      }
    }
    if (names.length > 0) return names.join('/');
  }
  return null;
}

export function parseProviderTeeLabel(rawLabel: string): ParsedTeeMetadata {
  const raw = rawLabel.trim();
  if (!raw) {
    return {
      rawLabel: raw,
      color: null,
      gender: null,
      isHistorical: false,
      isTemporary: false,
      isNineHoleHint: false,
      familyKey: '|',
    };
  }

  const lower = raw.toLowerCase();
  const isHistorical = HISTORICAL_TOKEN.test(raw) || (YEAR_SUFFIX.test(raw) && raw.includes(','));
  const isTemporary = lower.includes('temp');
  const isNineHoleHint = NINE_HOLE_HINT.test(raw);

  let color: string | null = null;
  let gender: 'men' | 'women' | null = null;

  if (!raw.includes(',')) {
    color = matchColorToken(raw) ?? titleCaseSingle(raw);
  } else {
    const parts = raw.split(',').map((p) => p.trim()).filter(Boolean);
    for (const part of parts) {
      const pl = part.toLowerCase();
      if (NUMERIC_COURSE_CODE.test(part) || RATING_SYSTEM.test(part)) continue;
      if (pl === 'men' || pl === 'man' || pl === 'male' || pl === 'mens') {
        gender = 'men';
        continue;
      }
      if (pl === 'women' || pl === 'woman' || pl === 'female' || pl === 'ladies' || pl === 'womens') {
        gender = 'women';
        continue;
      }
      const matched = matchColorToken(part);
      if (matched) color ??= matched;
    }
    color ??= matchColorToken(raw);
  }

  const familyKey = `${(color ?? raw.toLowerCase())}|${gender ?? ''}`;
  return { rawLabel: raw, color, gender, isHistorical, isTemporary, isNineHoleHint, familyKey };
}

export function teeDisplayLabel(meta: ParsedTeeMetadata, showGenderSuffix: boolean): string {
  const base = meta.color ?? titleCaseSingle(meta.rawLabel.split(',')[0]?.trim() ?? meta.rawLabel);
  if (showGenderSuffix && meta.gender) {
    return `${base} (${meta.gender === 'women' ? 'Women' : 'Men'})`;
  }
  return base;
}

export function teeColorHintFromMetadata(meta: ParsedTeeMetadata, existingHint: string | null): string | null {
  if (meta.color) return meta.color;
  const h = existingHint?.trim();
  return h && h.length > 0 ? h : null;
}

function teeHas18Distinct(holeNumbers: number[]): boolean {
  return new Set(holeNumbers).size >= 18;
}

export type TeeLike = {
  label: string;
  colorHint: string | null;
  holes: { holeNumber: number }[];
  yardageTotal: number;
};

export function preferTeeCandidate(
  incumbent: TeeLike,
  incumbentMeta: ParsedTeeMetadata,
  candidate: TeeLike,
  candidateMeta: ParsedTeeMetadata,
): boolean {
  const i18 = teeHas18Distinct(incumbent.holes.map((h) => h.holeNumber));
  const c18 = teeHas18Distinct(candidate.holes.map((h) => h.holeNumber));
  if (i18 !== c18) return c18;
  if (incumbentMeta.isHistorical !== candidateMeta.isHistorical) return !candidateMeta.isHistorical;
  if (incumbentMeta.isTemporary !== candidateMeta.isTemporary) return !candidateMeta.isTemporary;
  if (incumbentMeta.isNineHoleHint !== candidateMeta.isNineHoleHint) return !candidateMeta.isNineHoleHint;
  if (incumbent.yardageTotal !== candidate.yardageTotal) return candidate.yardageTotal > incumbent.yardageTotal;
  if (incumbent.holes.length !== candidate.holes.length) return candidate.holes.length > incumbent.holes.length;
  return candidateMeta.rawLabel.length < incumbentMeta.rawLabel.length;
}

/** Collapse provider tee list before DB upsert (same rules as Flutter prepareTeesForDisplay). */
export function filterProviderTeesForStorage<
  T extends { label: string; colorHint: string | null; holes: { holeNumber: number; yardageYds: number | null }[] },
>(tees: T[]): T[] {
  const nonempty = tees.filter((t) => t.holes.length > 0);
  if (nonempty.length === 0) return [];

  const metas = nonempty.map((t) => parseProviderTeeLabel(t.label));
  const byFamily = new Map<string, { tee: T; meta: ParsedTeeMetadata }>();

  for (let i = 0; i < nonempty.length; i++) {
    const tee = nonempty[i];
    const meta = metas[i];
    const yardageTotal = tee.holes.reduce((s, h) => s + (h.yardageYds ?? 0), 0);
    const prev = byFamily.get(meta.familyKey);
    if (!prev) {
      byFamily.set(meta.familyKey, { tee, meta });
      continue;
    }
    const prevY = prev.tee.holes.reduce((s, h) => s + (h.yardageYds ?? 0), 0);
    if (
      preferTeeCandidate(
        { label: prev.tee.label, colorHint: prev.tee.colorHint, holes: prev.tee.holes, yardageTotal: prevY },
        prev.meta,
        { label: tee.label, colorHint: tee.colorHint, holes: tee.holes, yardageTotal },
        meta,
      )
    ) {
      byFamily.set(meta.familyKey, { tee, meta });
    }
  }

  const genders = new Set(
    [...byFamily.values()].map((x) => x.meta.gender).filter((g): g is 'men' | 'women' => g != null),
  );
  const showGender = genders.size > 1;

  return [...byFamily.values()].map(({ tee, meta }) => ({
    ...tee,
    label: teeDisplayLabel(meta, showGender),
    colorHint: teeColorHintFromMetadata(meta, tee.colorHint),
  }));
}
